Add PositiveInteger scaler field

emZubair opened the pull request:
I want to merge this change because it adds scaler type `PositiveInteger` similar to `PositiveDecimal`. This can help remove the validations for input fields like `models.PositiveIntegerField`. 
Example:  i.e `ShippingMethod`'s `maximum_delivery_days` & `minimum_delivery_days` have validations in `clean_` methods to ensure we have +ve value, validations like this can be avoided with this. 

Suggestion: [There should be PositiveInteger similar to PositiveDecimal](https://github.com/saleor/saleor/discussions/15341)


# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

<!-- Docs are stored in a separate repository: https://github.com/saleor/saleor-docs/. -->
<!-- Please provide a link to the PR that updates documentation for your changes. -->
<!-- If changes in docs are not required, please mention that in the description. -->

- [ ] Link to documentation:

# Pull Request Checklist

<!-- Please keep this section. It will make the maintainer's life easier. -->

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/saleor/saleor?pullRequest=15342) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/saleor/saleor?pullRequest=15342) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/saleor/saleor?pullRequest=15342) it.</sub>

aniav commented:
Hey @emZubair could you add a changelog line please? Also the linter is complaining, could you make sure you run the pre-commit jobs to make sure it's all good there?

emZubair commented:
> Hey @emZubair could you add a changelog line please? Also the linter is complaining, could you make sure you run the pre-commit jobs to make sure it's all good there?

Hi @aniav I ran `black, isort and pylint` on my side. Can you tell me which linter should I run?
Regarding change log, I've added the PR description, I'm not sure what do you mean by changelog, I can't see any sample currently opened files, can you please share a sample? Do I need to update `changelog.md` file?

aniav commented:
@emZubair you can check the structure of the `CHANGELOG.md` file or read about changelog [in out docs](https://docs.saleor.io/docs/3.x/developer/community/contributing#changelog). The automatic linters we use should fire up automatically when you run `git commit` as pre-commit jobs. You can also trigger them manually with `bash .git/hooks/pre-commit`.

Out of curiosity, did the `pre-commit` hooks not fire up automatically for you? I noticed a few other PRs that seem to have got to us with linters complaining even though we have pre-commit setup to run on each commit.

emZubair commented:
> @emZubair you can check the structure of the `CHANGELOG.md` file or read about changelog [in out docs](https://docs.saleor.io/docs/3.x/developer/community/contributing#changelog). The automatic linters we use should fire up automatically when you run `git commit` as pre-commit jobs. You can also trigger them manually with `bash .git/hooks/pre-commit`.
> 
> Out of curiosity, did the `pre-commit` hooks not fire up automatically for you? I noticed a few other PRs that seem to have got to us with linters complaining even though we have pre-commit setup to run on each commit.

I've updated the PR and updated the `CHANGELOG.md`
No it didn't auto trigger the pre-commit hooks, I think those need to be installed manually before they start firing up. 
But I've updated the resolved the issues, you can re-trigger the jobs, it should be fine now.

aniav commented:
It's still broken on `mypy` checks. I do recommend installing the pre-commit hooks. It's a faster way around before the CI jobs do the same checks 🙌

emZubair commented:
> It's still broken on `mypy` checks. I do recommend installing the pre-commit hooks. It's a faster way around before the CI jobs do the same checks 🙌

The PR is updated.

emZubair commented:
> minimum_delivery_days

I can add a test case, but that will require me to update one of the existing mutations, to switch from `Int` to `PositiveInteger`

kadewu commented on saleor/graphql/core/scalars.py:
```diff
@@ -163,3 +164,19 @@ class Minute(graphene.Int):
     """The `Day` scalar type represents number of days by integer value."""
+
+
+class PositiveInteger(graphene.Int):
```
Should we follow the incorrect name convention with this new Scalar? 
By that I mean positive != nonnegative.

emZubair replied:
Why would you suggest? Considering Saleor is following the same naming convention for `PositiveDecimal` or there is a plan to refactor that as well?

mirekm replied:
Consistency > correctness in this case unless we're going to change it consistently across the board.

emZubair replied:
@mirekm Can you give some suggestion then, I have difficultly understanding how is the consistency being violated here? `PositiveInteger` or `PositiveInt`  are names that I can think of.

kadewu replied:
@emZubair My question was open question, not only to you, but other developers to start discussion here. It didn't mean that I disagree with proposed name.

mirekm replied:
@emZubair my comment was in support of your naming as it follows the current convention == consistent.

kadewu commented on saleor/graphql/core/scalars.py:
```diff
@@ -163,3 +164,19 @@ class Minute(graphene.Int):
+    """
+
+    @staticmethod
+    def parse_literal(ast):
```
Why do we only implement `parse_literal` method? We currently have almost similar problem with `PositiveDecimal` scalar (ref: https://github.com/saleor/saleor/issues/14373). It may be worth investigating if that's correct approach before we add new Scalar.

emZubair replied:
I the issue `serialize` method is mentioned, I couldn't find it in the inheritance hierarchy, from Int -> Scalar -> BaseType, UnmountedType -> OrderedType.
As far I've seen we need to implement custom serialize method in case of Decimal values, never needed one for Int types.

kadewu replied:
`serialize` and `parse_value` in `graphene.Int` have implementation, to be more exact, it's `coerce_int` method from `graphene.Int`

kadewu commented on saleor/graphql/core/scalars.py:
```diff
@@ -163,3 +164,19 @@ class Minute(graphene.Int):
+    def parse_literal(ast):
+        value = super(PositiveInteger, PositiveInteger).parse_literal(ast)
+        if value and value < 0:
+            raise GraphQLError(
```
We recently moved away from raising GraphQLError in scalars implementation. (ref: https://github.com/saleor/saleor/pull/14843)

emZubair replied:
`PositiveDecimal` returns `None`, should I do the same?

kadewu replied:
Yes.

kadewu reviewed (commented):
Is there a reason to add not used Scalar to the code?

emZubair commented:
> Is there a reason to add not used Scalar to the code?

We are using `PositiveIntegerField` and validations are placed to make sure -ve values is not entered, having `PositiveInteger` will make it easier and validations can be removed. 

I can close the PR, I thought it would be nice to have, but not a big issue 👍🏽

kadewu commented:
> > Is there a reason to add not used Scalar to the code?
> 
> We are using `PositiveIntegerField` and validations are placed to make sure -ve values is not entered, having `PositiveInteger` will make it easier and validations can be removed.
> 
> I can close the PR, I thought it would be nice to have, but not a big issue 👍🏽

It's another open question to other developers.
But in my opinion not used and not tested Scalar doesn't seem like a good idea.

aniav commented:
Hey @emZubair sorry for not reacting for some time. Thank you for your contribution.

In general, we shouldn't be adding any code that's not used anywhere aside from the definition. It makes it impossible to distinguish if a piece of unused code is just old or something we intend to use. This would also mean someone might just do cleanup in the codebase and removed it as it's unused.

I'll close this PR for now but feel free to reopen if you introduce the scalar to be used in any of the GraphQL types.
